src/Raven.Server/Documents/Replication/Senders/ReplicationDocumentSenderBase.cs

```diff
@@ -488,11 +492,37 @@ private bool AddReplicationItemToBatch(DocumentsOperationContext context, Replic
 
             if (item is AttachmentReplicationItem attachment)
             {
-                if (ShouldSendAttachmentStream(attachment))
-                    _replicaAttachmentStreams[attachment.Base64Hash] = attachment;
+                if (attachment.Flags != AttachmentFlags.Retired)
+                {
+                    if (ShouldSendAttachmentStream(attachment))
+                        _replicaAttachmentStreams[attachment.Base64Hash] = attachment;
 
-                if (MissingAttachmentsInLastBatch)
-                    state.MissingAttachmentBase64Hashes?.Remove(attachment.Base64Hash);
+                    if (MissingAttachmentsInLastBatch)
+                        state.MissingAttachmentBase64Hashes?.Remove(attachment.Base64Hash);
+                }
+                else if (_parent.Destination.Type != ReplicationNode.ReplicationType.Internal)
+                {
+                    if (ShouldSendAttachmentStream(attachment))
+                    {
+                        stats.RecordRetiredAttachmentStreamOutput(attachment.Size);
+                        //TODO: egor - I need a better way read teh stream, since this will fill the memory with attachment, the issue is I need to write the stream size first, and with network streams I can't do that,
+                        //so need to change the replication algorithm to allow for that
+                   
+                        var hash = attachment.Base64Hash.ToString();
+                        using Stream stream = Client.Util.AsyncHelpers.RunSync(() => _parent._database.DocumentsStorage.AttachmentsStorage.RetiredAttachmentsStorage.StreamForDownloadDestinationInternal(_downloader.Value, hash));
```

Comment:
The root cause of a lot of the problems is that you are trying to have retired attachments and eat the cake too.
You cannot have the same behavior and the same approach like that.

Separately - you can do a _separate_ call to get the length

Reply:
I need to unwrap the attachment in external replication, but I agree there might be better solution such as reading it into a network stream directly then saving it into a memory stream,